IQSS/dataverse-frontend
# SPA - OIDC integration following the design implemented on the PoC

## What this PR does / why we need it:
Change SPA authentication method to use Keycloak as OIDC provider using PKCE based on an [initial PoC](https://github.com/IQSS/dataverse-frontend/pull/516).


## Which issue(s) this PR closes:

- Closes #528 

## Special notes for your reviewer:
This PR does not point to the `develop` branch but to a separate branch ( [authentication-oidc](https://github.com/IQSS/dataverse-frontend/tree/authentication-oidc) ), the idea is in another PR to configure the deployment to another environment that is not Beta maybe to be able to test this not only locally before merging to develop.

All unit tests and e2e tests are working now with this new authentication method.
Application terms of use are empty for now as we don't have an API endpoint for it yet.

## Suggestions on how to test this:

**Step 1: Run the Development Environment**
1. Execute `npm i`.
2. Navigate with `cd packages/design-system && npm i && npm run build`.
3. Return with `cd ../../`.
4. Ensure you have a `.env` file similar to `.env.example`, with the variable `VITE_DATAVERSE_BACKEND_URL=http://localhost:8000`.
5. Make sure your `dev-env/.env` file is filled with the appropriate data. For the `REGISTRY` variable, please set `ghcr.io`, as we will use a [PR-generated](https://github.com/IQSS/dataverse/pull/10972) dataverse image.
6. Navigate with `cd dev-env`.
7. Start the environment using `./run-env.sh 10959-bearer-token-auth-ext `.
8. To verify the environment, visit <http://localhost:8000> and check your local Dataverse installation.

**Step 2: Test the feature in the SPA**
1. Click the Log In button in the header.
2. You will be redirected to the Keycloak login page
3. Enter the credentials, there are 4 pre-set users in keycloak, you can log in with any of those users:

- dataverse-user@mailinator.com -  **_pass_**: user
- dataverse-admin@mailinator.com - **_pass_**: admin
- dataverse-affiliate@mailinator.com - **_pass_**: affiliate
- dataverse-curator@mailinator.com - **_pass_**: curator

4. After the first login, you will be redirected to finish the registration process, where you will have to accept the terms (although there are none for now) and you will also be able to optionally add the affiliation and position. 
5.  After submitting the form, you will be redirected to the “root” collection page with a welcome success alert and some confetti. 🎉 .
6. Click on the user drop-down menu in the header and go to Account Information, check that the data displayed there matches the data you have entered during the registration process.
7. Try creating a collection and a dataset as a smoke test.
8. Now log out and log in again, the registration process should not be required now 👍🏼 

Another path:
1. Follow the first 4 steps as before (but with a different user that didn't complete the registration process yet), but now click the Cancel button in the Sign Up page, you should be able to navigate through the SPA as an unauthenticated user.

## Does this PR introduce a user interface change? If mockups are available, please link/include them here:
![sign-up-page](https://github.com/user-attachments/assets/dc8694d7-4d71-4633-8738-dd2e264f92e7)

https://github.com/user-attachments/assets/99a209a9-9f08-40d4-ac6a-bfbbb3e005fd

## Is there a release notes update needed for this change?: 
No

## Additional documentation:
No
